projects/seeker/scripts/generate_lm_data.py

```diff
@@ -0,0 +1,346 @@
+            '--f1-overlap-threshold',
+            type=float,
+            default=0.5,
+            help='Threshold of f1 in order to keep examples',
+        )
+        group.add_argument(
+            '--min-num-search-words',
+            type=float,
+            default=-1,
+            help='Threshold of how many words required in search term to search example (exclusive)',
+        )
+        group.add_argument(
+            '--write-every-n-valid-exs',
+            type=float,
+            default=1000,
+            help='How many valid exs to write for each log',
+        )
+        parser.set_defaults(rag_retriever_type=RetrieverType.SEARCH_ENGINE.value)
+        return parser
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt)
+        if shared:
+            self.search_engine = shared['search_engine']
+        else:
+            self.search_engine = OverlapSearchEngine(
+                opt, DictionaryAgent(opt), shared=shared
+            )  # type: ignore
+        self.threshold = opt['f1_overlap_threshold']
+        self.dummy = torch.zeros(1, 1, dtype=torch.long)
+
+    def share(self):
+        shared = super().share()
+        shared['search_engine'] = self.search_engine
+        return shared
+
+    def construct_search_query(self, labels: List[str]) -> List[str]:
+        """
+        Construct the search query
+
+        :param observation:
+            observation from task
+
+        :return query:
+            return search query.
+        """
+        assert labels
+        search_query = [
+            ' '.join(
+                [
+                    w
+                    for w in DictionaryAgent.split_tokenize(labels[0])
+                    if not filter_word(w)
+                ]
+            )
+        ]
+        return search_query
+
+    def get_best_doc(
+        self, all_docs: List[Document], labels: List[str]
+    ) -> Tuple[Optional[float], Optional[Document], Optional[int]]:
+        """
+        Given a set of all retrieved docs, determine best fitting document.
+
+        :param all_docs:
+            list of all retrieved Documents
+        :param labels:
+            labels for the current example
+
+        :return (best_f1, best_doc, best_doc_idx):
+            return the best document, along with the f1 overlap and index into all_docs
+
+        """
+        docs = []
+        for i, d in enumerate(all_docs):
+            if d.startswith('.'):
+                d = d[2:]
+            try:
+                docs += [(i, s) for s in nltk.sent_tokenize(d)]
+            except IndexError:
+                # Something's up with the NLTK Sentence tokenizer here.
+                docs += [(i, s) for s in d.split('.')]
+        f1s, inds = torch.FloatTensor(
+            [F1Metric.compute(labels[0], [d]).value() for _, d in docs]
+        ).topk(len(docs))
+        best_doc = None
+        best_doc_idx = None
+        best_f1 = None
+        for f1, ind in zip(f1s, inds):
```

Comment:
The function name says best doc. Why does it looks like it picks the first doc that is above the threshold?

Reply:
ahh because we assume that the docs are retrieved in order of their quality. so presumably the first doc that is above the threshold is in fact the best doc

Reply:
They are, but the metric used by the search engine might be different by the F1 metric we use here.

Reply:
agreed, we're also placing our trust more so in the search engine than in a heuristic word overlap metric